### PR TITLE
Feature/dsnd 612 - Return non-2XX response codes - Metrics API

### PR DESCRIPTION
**Pull request by ipg-ch**

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/company/metrics/exception/ControllerExceptionHandler.java [hunk 1]
```diff
@@ -46,9 +57,43 @@ public ResponseEntity<Object> handleException(Exception ex, WebRequest request)
         responseBody.put("correlationId", correlationId);
         request.setAttribute("javax.servlet.error.exception", ex, 0);
         logger.error("correlationId = " + correlationId, ex);
+
         return new ResponseEntity<>(responseBody, HttpStatus.INTERNAL_SERVER_ERROR);
     }
 
+    @ExceptionHandler(value = {DataAccessResourceFailureException.class})
+    public ResponseEntity<Object> handleException(DataAccessResourceFailureException ex, WebRequest request) {
```
Comment (iichr) [inline src/main/java/uk/gov/companieshouse/company/metrics/exception/ControllerExceptionHandler.java:65]: [non-blocking] Seems like the request argument in all of these handlers is redundant

File: src/main/java/uk/gov/companieshouse/company/metrics/exception/ControllerExceptionHandler.java [hunk 2]
```diff
@@ -46,9 +57,43 @@ public ResponseEntity<Object> handleException(Exception ex, WebRequest request)
         responseBody.put("correlationId", correlationId);
         request.setAttribute("javax.servlet.error.exception", ex, 0);
         logger.error("correlationId = " + correlationId, ex);
+
         return new ResponseEntity<>(responseBody, HttpStatus.INTERNAL_SERVER_ERROR);
     }
 
+    @ExceptionHandler(value = {DataAccessResourceFailureException.class})
```
Comment (iichr) [inline src/main/java/uk/gov/companieshouse/company/metrics/exception/ControllerExceptionHandler.java:64]: [non-blocking] Elsewhere we've gone for the whole parent class of exceptions `DataAccessException`, though it is true that manually brining down the DB after starting the app produces the one in your handler. Would this be sufficient however to catch all possible cases when we'd like to throw service unavailable?

**Review by iichr — COMMENTED** [review]
The unit tests for the controller advice class in isolation are nice to have, but would be good to see some tests for both the controller and service - will prove the correct exception handler is kicking in.

For inspiration, we have done this in the following 2 PRs so hopefully it won't be too much effort to lift and shift:
- https://github.com/companieshouse/insolvency-data-api/pull/28
- https://github.com/companieshouse/company-profile-api/pull/28

File: src/main/java/uk/gov/companieshouse/company/metrics/exception/ControllerExceptionHandler.java [hunk 3]
```diff
@@ -46,9 +57,43 @@ public ResponseEntity<Object> handleException(Exception ex, WebRequest request)
...
+        logger.error(String.format("Started: handleException: %s Generating error response ",
+                correlationId), ex);
+        Map<String, Object> responseBody = new LinkedHashMap<>();
+        populateResponseBody(responseBody, correlationId);
+        return new ResponseEntity<>(responseBody, HttpStatus.BAD_REQUEST);
+    }
+
     private String generateShortCorrelationId() {
         return UUID.randomUUID().toString().replace(TARGET, REPLACEMENT).substring(0,8);
     }
```
[the remark this answers is not in the document]
Comment (ipg-ch) [reply src/main/java/uk/gov/companieshouse/company/metrics/exception/ControllerExceptionHandler.java:99]: Was told this was required. Will verify.

File: src/main/java/uk/gov/companieshouse/company/metrics/exception/ControllerExceptionHandler.java (hunk 1 above)
Comment (ipg-ch) [reply src/main/java/uk/gov/companieshouse/company/metrics/exception/ControllerExceptionHandler.java:65]: Missed request.setAttribute("javax.servlet.error.exception", ex, 0);

File: src/main/java/uk/gov/companieshouse/company/metrics/exception/ControllerExceptionHandler.java (hunk 2 above)
Comment (ipg-ch) [reply src/main/java/uk/gov/companieshouse/company/metrics/exception/ControllerExceptionHandler.java:64]: Updated handler, unit test still passes.

File: src/main/java/uk/gov/companieshouse/company/metrics/exception/ControllerExceptionHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ipg-ch) [reply src/main/java/uk/gov/companieshouse/company/metrics/exception/ControllerExceptionHandler.java:102]: Confirmed with Tech Lead is required.

**Pull request merged**